File: guests/swift/panes.swift
// The panes scene, Swift port — guests/rust/panes.rs,
// tools/scenes/panes.steps.

import Foundation
import Kaya

let CONTENT: UInt64 = 7
let DETAIL: UInt64 = 8

let app = KayaApp()

app.build { tx in
    tx.window(title: "panes", panes: 3)
    let root = tx.column { root in
        // Authored ids: an index read passes for an arm that drew nothing.
        let caption = tx.signal(.str("root pane"))
        tx.setA11yId(tx.label(bind: caption), "root")  // label#0
        tx.button(
            "open content",
            onClick: { content in  // button#0
                content.pushEntry(CONTENT, title: "content")
                let pane = content.column { pane in
                    let caption = content.signal(.str("content pane"))
                    content.setA11yId(content.label(bind: caption), "content")  // label#1
                    content.button(
                        "open detail",
                        onClick: { detail in  // button#1
                            detail.pushEntry(DETAIL, title: "detail")
                            let pane = detail.column { pane in
                                let caption = detail.signal(.str("detail pane"))
                                // label#last
                                detail.setA11yId(detail.label(bind: caption), "detail")
                                return pane
                            }
                            detail.mountIn(DETAIL, pane)
                        })
                    return pane
                }
                content.mountIn(CONTENT, pane)
            })
        return root
    }
    tx.mount(root)
}

app.run()

File: guests/ocaml/panes.ml
(* The panes scene, OCaml port — guests/rust/panes.rs,
   tools/scenes/panes.steps. *)

open Kaya_app

let content = 7L
let detail = 8L

let () =
  let app = Kaya_app.create () in

  build app (fun () ->
     window ~title:"panes" ~panes:3 ();

     let on_detail () =
       push_entry ~title:"detail" detail;
       let caption = signal Scalar.Str ("detail pane") in
       let pane = column [ label ~a11y_id:"detail" ~bind:caption ] () in
       mount_in detail pane
     in
     let on_content () =
       push_entry ~title:"content" content;
       let caption = signal Scalar.Str ("content pane") in
       let pane =
         column
           [
             label ~a11y_id:"content" ~bind:caption (* label#1 *);
             button ~text:"open detail" ~on_click:on_detail (* button#1 *);
           ]
           ()
       in
       mount_in content pane
     in

     let caption = signal Scalar.Str ("root pane") in
     let root =
       column
         [
           (* Authored ids: an index read passes whether or not anything
              reached the screen. *)
           label ~a11y_id:"root" ~bind:caption (* label#0 *);
           button ~text:"open content" ~on_click:on_content (* button#0 *);
         ]
         ()
     in
     mount root);

  exit (run app)
